File: Java/Decirational/src/Rational.java
package decirational;

import java.util.HashMap;
import java.util.Objects;
import java.util.function.Function;

public final class Rational<T extends CustomInteger<T>> implements Comparable<Rational<T>> {
    private final T numerator;
    private final T denominator;

    public T getNumerator() {
        return numerator;
    }

    public T getDenominator() {
        return denominator;
    }

    private Rational<T> getOne() {
        final T one = denominator.pow(0);
        return new Rational<>(one, one, true, true, true);
    }

    private T getFive() {
        final T one = denominator.pow(0);
        final T two = one.plus(one);
        final T four = two.plus(two);
        return four.plus(one);
    }

    private T getTen() {
        final T one = denominator.pow(0);
        final T two = one.plus(one);
        final T four = two.plus(two);
        return four.multiply(two).plus(two);
    }

    public Rational(T numerator, T denominator) {
        if (null == numerator || null == denominator) {
            throw new NullPointerException("numerator and denominator cannot be null");
        }
        if (denominator.isZero()) {
            throw new IllegalArgumentException("denominator cannot be zero");
        }
        final T gcd = numerator.gcd(denominator);
        if (denominator.isNegative()) {
            numerator = numerator.negate();
            denominator = denominator.negate();
        }
        this.numerator = numerator.divideBy(gcd);
        this.denominator = denominator.divideBy(gcd);
    }

    private Rational(T numerator, T denominator, final boolean input_unsafe) {
        final T gcd = numerator.gcd(denominator);
        if (denominator.isNegative()) {
            numerator = numerator.negate();
            denominator = denominator.negate();
        }
        this.numerator = numerator.divideBy(gcd);
        this.denominator = denominator.divideBy(gcd);
    }

    private Rational(T numerator, T denominator, final boolean input_unsafe, final boolean gcd_unsafe) {
        if (denominator.isNegative()) {
            numerator = numerator.negate();
            denominator = denominator.negate();
        }
        this.numerator = numerator;
        this.denominator = denominator;
    }

    private Rational(final T numerator, final T denominator, final boolean input_unsafe, final boolean gcd_unsafe, final boolean sign_unsafe) {
        this.numerator = numerator;
        this.denominator = denominator;
    }

    public Rational(final T integer) {
        if (null == integer) {
            throw new NullPointerException("integer cannot be null");
        }
        this.numerator = integer;
        this.denominator = integer.pow(0);
    }

    private Rational(final T integer, final boolean input_unsafe) {
        this.numerator = integer;
        this.denominator = integer.pow(0);
    }

    public Rational(final Rational<T> rational) {
        if (null == rational) {
            throw new NullPointerException("rational cannot be null");
        }
        this.numerator = rational.numerator;
        this.denominator = rational.denominator;
    }

    public Rational(String string, final Function<String, T> from_string) {
        if (null == string) {
            throw new IllegalArgumentException("input is null");
        }
        if (null == from_string) {
            throw new IllegalArgumentException("integer constructor is null");
        }
        string = string.replaceAll("\\s", "");
        if (string.isEmpty()) {
            throw new NumberFormatException("input is empty");
        }
        boolean negative = false;
        int starting_point = 0;
        if (Arithmetic.isMinus(string.charAt(0))) {
            negative = true;
            starting_point = 1;
        } else if (Arithmetic.isPlus(string.charAt(0))) {
            starting_point = 1;
        }
        if (starting_point >= string.length() || !Arithmetic.isDigit(string.charAt(starting_point))) {
            throw new NumberFormatException("the rational string does not have the right format");
        }
        int fraction_bar = -1;
        int decimal_point = -1;
        int cyclic_begin = -1;
        int cyclic_end = -1;
        for (int i = starting_point + 1; i < string.length(); ++i) {
            if (Arithmetic.isFractionBar(string.charAt(i))) {
                if (-1 != fraction_bar || -1 != decimal_point) {
                    throw new NumberFormatException("the rational string does not have the right format");
                }
                fraction_bar = i;
            } else if (Arithmetic.isDecimalPoint(string.charAt(i))) {
                if (-1 != decimal_point || -1 != fraction_bar || i == string.length() - 1) {
                    throw new NumberFormatException("the rational string does not have the right format");
                }
                decimal_point = i;
            } else if (Arithmetic.isCyclicBegin(string.charAt(i))) {
                if (-1 != cyclic_begin || -1 == decimal_point) {
                    throw new NumberFormatException("the rational string does not have the right format");
                }
                cyclic_begin = i;
            } else if (Arithmetic.isCyclicEnd(string.charAt(i))) {
                if (-1 == cyclic_begin || -1 != cyclic_end || i != string.length() - 1 || 1 == i - cyclic_begin) {
                    throw new NumberFormatException("the rational string does not have the right format");
                }
                cyclic_end = i;
            } else if (!Arithmetic.isDigit(string.charAt(i))) {
                throw new NumberFormatException("the rational string does not have the right format");
            }
        }
        if (-1 == decimal_point) {
            final String numerator_string, denominator_string;
            if (-1 == fraction_bar) {
                numerator_string = string.substring(starting_point);
                denominator_string = "1";
            } else if (fraction_bar != string.length() - 1) {
                numerator_string = string.substring(starting_point, fraction_bar);
                denominator_string = string.substring(fraction_bar + 1);
            } else {
                throw new NumberFormatException("the rational string does not have the right format");
            }
            T numerator;
            final T denominator;
            try {
                numerator = from_string.apply(numerator_string);
                denominator = from_string.apply(denominator_string);
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("cannot instantiate a rational from the given integer type", e);
            }
            if (denominator.isZero()) {
                throw new IllegalArgumentException("denominator cannot be zero");
            }
            final T gcd = numerator.gcd(denominator);
            if (negative) {
                numerator = numerator.negate();
            }
            this.numerator = numerator.divideBy(gcd);
            this.denominator = denominator.divideBy(gcd);
        } else if (-1 == cyclic_begin) {
            final String numerator_string = string.substring(starting_point, decimal_point) + string.substring(decimal_point + 1);
            final String denominator_string = "1" + "0".repeat(string.length() - decimal_point - 1);
            T numerator;
            final T denominator;
            try {
                numerator = from_string.apply(numerator_string);
                denominator = from_string.apply(denominator_string);
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("cannot instantiate a rational from the given integer type", e);
            }
            final T gcd = numerator.gcd(denominator);
            if (negative) {
                numerator = numerator.negate();
            }
            this.numerator = numerator.divideBy(gcd);
            this.denominator = denominator.divideBy(gcd);
        } else if (-1 != cyclic_end) {
            final String finite_numerator_string = string.substring(starting_point, decimal_point) + string.substring(decimal_point + 1, cyclic_begin);
            final String finite_denominator_string = "1" + "0".repeat(cyclic_begin - decimal_point - 1);
            final String cyclic_numerator_string = string.substring(cyclic_begin + 1, cyclic_end);
            final String cyclic_denominator_string = "9".repeat(cyclic_end - cyclic_begin - 1) + "0".repeat(cyclic_begin - decimal_point - 1);
            final T finite_numerator;
            final T finite_denominator;
            final T cyclic_numerator;
            final T cyclic_denominator;
            try {
                finite_numerator = from_string.apply(finite_numerator_string);
                finite_denominator = from_string.apply(finite_denominator_string);
                cyclic_numerator = from_string.apply(cyclic_numerator_string);
                cyclic_denominator = from_string.apply(cyclic_denominator_string);
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("cannot instantiate a rational from the given integer type", e);
            }
            final Rational<T> finite = new Rational<>(finite_numerator, finite_denominator, true);
            final Rational<T> cyclic = new Rational<>(cyclic_numerator, cyclic_denominator, true);
            Rational<T> result = finite.plus(cyclic);
            if (negative) {
                result = result.negate();
            }
            this.numerator = result.numerator;
            this.denominator = result.denominator;
        } else {
            throw new NumberFormatException("the rational string does not have the right format");
        }
    }

    @Override
    public String toString() {
        if (denominator.isOne()) {
            return numerator.toString();
        }
        return toFractionString();
    }

    public String toFractionString() {
        return numerator + "/" + denominator;
    }

    public String toMixedString() {
        if (denominator.isOne()) {
            return numerator.toString();
        }
        final T[] whole_and_remainder = numerator.abs().divideByAndModulo(denominator);
        final T whole = whole_and_remainder[0];
        final T remainder = whole_and_remainder[1];
        final StringBuilder sb = new StringBuilder();
        if (numerator.isNegative()) {
            sb.append('-');
        }
        if (!whole.isZero()) {
            sb.append(whole).append(' ');
        }
        sb.append(remainder).append('/').append(denominator);
        return sb.toString();
    }

    public String toDecimalString() {
        final T ten = getTen();
        T[] integer_and_remainder = numerator.abs().divideByAndModulo(denominator);
        final T whole_integer = integer_and_remainder[0];
        T remainder = integer_and_remainder[1];
        final StringBuilder decimal = new StringBuilder();
        decimal.append(numerator.isNegative() ? '-' : "").append(whole_integer).append(remainder.isZero() ? "" : '.');
        final HashMap<T, Integer> remainder_map = new HashMap<>();
        remainder_map.put(remainder, decimal.length());
        int starting_cyclic = -1;
        while (!remainder.isZero() && -1 == starting_cyclic) {
            remainder = remainder.multiply(ten);
            integer_and_remainder = remainder.divideByAndModulo(denominator);
            final T integer = integer_and_remainder[0];
            remainder = integer_and_remainder[1];
            decimal.append(integer.toString().charAt(0));
            if (remainder_map.containsKey(remainder)) {
                starting_cyclic = remainder_map.get(remainder);
            }
            remainder_map.put(remainder, decimal.length());
        }
        if (-1 != starting_cyclic) {
            decimal.insert(starting_cyclic, Arithmetic.cyclic_begin);
            decimal.append(Arithmetic.cyclic_end);
        }
        return decimal.toString();
    }

    public String toTruncateDecimalString(final int round_to) {
        final T ten = getTen();
        T[] integer_and_remainder = numerator.abs().divideByAndModulo(denominator);
        final T whole_integer = integer_and_remainder[0];
        final String whole_integer_string = whole_integer.toString();
        if (whole_integer_string.length() <= -round_to) {
            return (numerator.isNegative() ? '-' : "") + "0";
        } else if (round_to < 0) {
            if (Integer.MIN_VALUE == round_to) {
                throw new IllegalArgumentException("cannot round to the minimum representable precision");
            }
            final T shift_base = ten.pow(-round_to);
            final T result = whole_integer.divideBy(shift_base).multiply(shift_base);
            return (numerator.isNegative() ? '-' : "") + result.toString();
        } else if (0 == round_to) {
            return (numerator.isNegative() ? '-' : "") + whole_integer_string;
        }
        T remainder = integer_and_remainder[1];
        final StringBuilder decimal = new StringBuilder();
        decimal.append(numerator.isNegative() ? '-' : "").append(whole_integer_string).append(remainder.isZero() ? "" : '.');
        int index = 0;
        while (index < round_to && !remainder.isZero()) {
            remainder = remainder.multiply(ten);
            integer_and_remainder = remainder.divideByAndModulo(denominator);
            final T integer = integer_and_remainder[0];
            remainder = integer_and_remainder[1];
            decimal.append(integer.toString().charAt(0));
            ++index;
        }
        return decimal.toString();
    }

    public String toRoundDecimalString(final int round_to) {
        if (Integer.MIN_VALUE == round_to) {
            throw new IllegalArgumentException("cannot round to the minimum representable precision");
        }
        final Rational<T> five = new Rational<>(getFive(), true);
        final Rational<T> ten = new Rational<>(getTen(), true);
        final Rational<T> shift_base = ten.pow(-round_to - 1);
        final Rational<T> delta = five.multiply(shift_base);
        final Rational<T> temp_rational = isNegative() ? minus(delta) : plus(delta);
        return temp_rational.toTruncateDecimalString(round_to);
    }

    public String toCeilDecimalString(final int round_to) {
        if (isNegative()) {
            return '-' + negate().toFloorDecimalString(round_to);
        }
        if (Integer.MIN_VALUE == round_to) {
            throw new IllegalArgumentException("cannot round to the minimum representable precision");
        }
        final Rational<T> ten = new Rational<>(getTen(), true);
        if (multiply(ten.pow(round_to)).isInteger()) {
            return toTruncateDecimalString(round_to);
        }
        final Rational<T> shift_base = ten.pow(-round_to);
        final Rational<T> temp_rational = plus(shift_base);
        return temp_rational.toTruncateDecimalString(round_to);
    }

    public String toFloorDecimalString(final int round_to) {
        if (isNegative()) {
            return '-' + negate().toCeilDecimalString(round_to);
        }
        return toTruncateDecimalString(round_to);
    }

    public boolean isInteger() {
        return denominator.isOne();
    }

    public boolean isZero() {
        return numerator.isZero();
    }

    public boolean isPositive() {
        return numerator.isPositive();
    }

    public boolean isNegative() {
        return numerator.isNegative();
    }

    public Rational<T> negate() {
        return new Rational<>(numerator.negate(), denominator, true, true, true);
    }

    public Rational<T> abs() {
        return new Rational<>(numerator.abs(), denominator, true, true, true);
    }

    public Rational<T> reciprocal() {
        if (isZero()) {
            throw new ArithmeticException("cannot get the reciprocal of zero");
        }
        return new Rational<>(denominator, numerator, true, true);
    }

    private Rational<T> reciprocal(final boolean unsafe) {
        return new Rational<>(denominator, numerator, true, true);
    }

    public T getNumeratorAbs() {
        return numerator.abs();
    }

    public T getDenominatorAbs() {
        return denominator.abs();
    }

    @Override
    public int compareTo(final Rational<T> other) {
        final T gcd = denominator.gcd(other.denominator);
        return numerator.multiply(other.denominator.divideBy(gcd)).compareTo(other.numerator.multiply(denominator.divideBy(gcd)));
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final Rational<?> other = (Rational<?>) o;
        return numerator.equals(other.numerator) && denominator.equals(other.denominator);
    }

    @Override
    public int hashCode() {
        return Objects.hash(numerator, denominator);
    }

    public Rational<T> plus(final Rational<T> other) {
        final T gcd = this.denominator.gcd(other.denominator);
        final T multiplier_1 = other.denominator.divideBy(gcd);
        final T multiplier_2 = this.denominator.divideBy(gcd);
        final T numerator = this.numerator.multiply(multiplier_1).plus(other.numerator.multiply(multiplier_2));
        final T denominator = gcd.multiply(multiplier_1).multiply(multiplier_2);
        return new Rational<>(numerator, denominator, true);
    }

    public Rational<T> minus(final Rational<T> other) {
        final T gcd = this.denominator.gcd(other.denominator);
        final T multiplier_1 = other.denominator.divideBy(gcd);
        final T multiplier_2 = this.denominator.divideBy(gcd);
        final T numerator = this.numerator.multiply(multiplier_1).minus(other.numerator.multiply(multiplier_2));
        final T denominator = gcd.multiply(multiplier_1).multiply(multiplier_2);
        return new Rational<>(numerator, denominator, true);
    }

    public Rational<T> multiply(final Rational<T> other) {
        final T gcd_1 = this.denominator.gcd(other.numerator);
        final T gcd_2 = other.denominator.gcd(this.numerator);
        final T denominator = this.denominator.divideBy(gcd_1).multiply(other.denominator.divideBy(gcd_2));
        final T numerator = this.numerator.divideBy(gcd_2).multiply(other.numerator.divideBy(gcd_1));
        return new Rational<>(numerator, denominator, true, true, true);
    }

    public Rational<T> divideBy(final Rational<T> other) {
        if (other.isZero()) {
            throw new ArithmeticException("cannot divide by zero");
        }
        return multiply(other.reciprocal(true));
    }

    public Rational<T> pow(final int exponent) {
        if (exponent < 0) {
            if (isZero()) {
                throw new ArithmeticException("exponent cannot be negative for zero");
            }
            if (Integer.MIN_VALUE == exponent) {
                return reciprocal(true).pow(Integer.MAX_VALUE).multiply(reciprocal(true));
            }
            return reciprocal(true).pow(-exponent);
        } else if (0 == exponent) {
            return getOne();
        } else if (1 == exponent) {
            return this;
        }
        Rational<T> result = getOne();
        Rational<T> base = this;
        int power = exponent;
        while (true) {
            if (1 == (power & 1)) {
                result = result.multiply(base);
            }
            power >>= 1;
            if (power > 0) {
                base = base.multiply(base);
            } else {
                break;
            }
        }
        return result;
    }
}

File: Go/Decirational/rational.go
package decirational

import (
	"fmt"
	"math"
	"strings"
)

// Rational[T] is an exact fraction over a CustomInteger backend.
type Rational[T CustomInteger[T]] struct {
	numerator   T
	denominator T
}

func (r Rational[T]) GetNumerator() T   { return r.numerator }
func (r Rational[T]) GetDenominator() T { return r.denominator }

func (r Rational[T]) getOne() Rational[T] {
	one := r.denominator.Pow(0)
	return Rational[T]{one, one}
}

func (r Rational[T]) getFive() T {
	one := r.denominator.Pow(0)
	two := one.Plus(one)
	four := two.Plus(two)
	return four.Plus(one)
}

func (r Rational[T]) getTen() T {
	one := r.denominator.Pow(0)
	two := one.Plus(one)
	four := two.Plus(two)
	return four.Multiply(two).Plus(two)
}

func rationalReduced[T CustomInteger[T]](numerator, denominator T) Rational[T] {
	gcd := numerator.Gcd(denominator)
	if denominator.IsNegative() {
		numerator = numerator.Negate()
		denominator = denominator.Negate()
	}
	return Rational[T]{numerator.DivideBy(gcd), denominator.DivideBy(gcd)}
}

func rationalSignNormalized[T CustomInteger[T]](numerator, denominator T) Rational[T] {
	if denominator.IsNegative() {
		numerator = numerator.Negate()
		denominator = denominator.Negate()
	}
	return Rational[T]{numerator, denominator}
}

// NewRational builds a reduced, sign-normalized fraction, rejecting a zero denominator.
func NewRational[T CustomInteger[T]](numerator, denominator T) (Rational[T], error) {
	if denominator.IsZero() {
		return Rational[T]{}, fmt.Errorf("denominator cannot be zero")
	}
	return rationalReduced(numerator, denominator), nil
}

// NewRationalFromInteger builds the fraction integer/1.
func NewRationalFromInteger[T CustomInteger[T]](integer T) Rational[T] {
	return Rational[T]{integer, integer.Pow(0)}
}

// ParseRational parses a fraction ("3/4"), decimal ("0.5"), or repeating-decimal
// ("0.{3}", "1.5{6}") literal into a reduced Rational[T]. parseInt builds a T
// from a plain (unsigned, digits-only) decimal string - supplied explicitly
// since a bare CustomInteger[T] constraint gives no way to require one.
func ParseRational[T CustomInteger[T]](s string, parseInt func(string) (T, error)) (Rational[T], error) {
	s = stripWhitespace(s)
	if s == "" {
		return Rational[T]{}, fmt.Errorf("input is empty")
	}
	negative := false
	start := 0
	if isMinus(s[0]) {
		negative = true
		start = 1
	} else if isPlus(s[0]) {
		start = 1
	}
	if start >= len(s) || !isDigit(s[start]) {
		return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
	}
	fractionBar, decimalPoint, cyclicStart, cyclicEnd := -1, -1, -1, -1
	for i := start + 1; i < len(s); i++ {
		switch {
		case isFractionBar(s[i]):
			if fractionBar != -1 || decimalPoint != -1 {
				return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
			}
			fractionBar = i
		case isDecimalPoint(s[i]):
			if decimalPoint != -1 || fractionBar != -1 || i == len(s)-1 {
				return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
			}
			decimalPoint = i
		case isCyclicBeginChar(s[i]):
			if cyclicStart != -1 || decimalPoint == -1 {
				return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
			}
			cyclicStart = i
		case isCyclicEndChar(s[i]):
			if cyclicStart == -1 || cyclicEnd != -1 || i != len(s)-1 || i-cyclicStart == 1 {
				return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
			}
			cyclicEnd = i
		case !isDigit(s[i]):
			return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
		}
	}

	if decimalPoint == -1 {
		var numeratorStr, denominatorStr string
		if fractionBar == -1 {
			numeratorStr, denominatorStr = s[start:], "1"
		} else if fractionBar != len(s)-1 {
			numeratorStr, denominatorStr = s[start:fractionBar], s[fractionBar+1:]
		} else {
			return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
		}
		numerator, err := parseInt(numeratorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		denominator, err := parseInt(denominatorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		if denominator.IsZero() {
			return Rational[T]{}, fmt.Errorf("denominator cannot be zero")
		}
		if negative {
			numerator = numerator.Negate()
		}
		return rationalReduced(numerator, denominator), nil
	} else if cyclicStart == -1 {
		numeratorStr := s[start:decimalPoint] + s[decimalPoint+1:]
		denominatorStr := "1" + strings.Repeat("0", len(s)-decimalPoint-1)
		numerator, err := parseInt(numeratorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		denominator, err := parseInt(denominatorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		if negative {
			numerator = numerator.Negate()
		}
		return rationalReduced(numerator, denominator), nil
	} else if cyclicEnd != -1 {
		finiteNumeratorStr := s[start:decimalPoint] + s[decimalPoint+1:cyclicStart]
		finiteDenominatorStr := "1" + strings.Repeat("0", cyclicStart-decimalPoint-1)
		cyclicNumeratorStr := s[cyclicStart+1 : cyclicEnd]
		cyclicDenominatorStr := strings.Repeat("9", cyclicEnd-cyclicStart-1) + strings.Repeat("0", cyclicStart-decimalPoint-1)
		finiteNumerator, err := parseInt(finiteNumeratorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		finiteDenominator, err := parseInt(finiteDenominatorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		cyclicNumerator, err := parseInt(cyclicNumeratorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		cyclicDenominator, err := parseInt(cyclicDenominatorStr)
		if err != nil {
			return Rational[T]{}, fmt.Errorf("cannot instantiate a rational from the given integer type: %w", err)
		}
		finite := rationalReduced(finiteNumerator, finiteDenominator)
		cyclic := rationalReduced(cyclicNumerator, cyclicDenominator)
		result := finite.Plus(cyclic)
		if negative {
			result = result.Negate()
		}
		return result, nil
	}
	return Rational[T]{}, fmt.Errorf("the rational string does not have the right format")
}

func (r Rational[T]) String() string {
	if r.denominator.IsOne() {
		return r.numerator.String()
	}
	return r.ToFractionString()
}

func (r Rational[T]) ToFractionString() string {
	return r.numerator.String() + "/" + r.denominator.String()
}

func (r Rational[T]) ToMixedString() string {
	if r.denominator.IsOne() {
		return r.numerator.String()
	}
	wr := r.numerator.Abs().DivideByAndModulo(r.denominator)
	whole, remainder := wr[0], wr[1]
	var sb strings.Builder
	if r.numerator.IsNegative() {
		sb.WriteByte('-')
	}
	if !whole.IsZero() {
		sb.WriteString(whole.String())
		sb.WriteByte(' ')
	}
	sb.WriteString(remainder.String())
	sb.WriteByte('/')
	sb.WriteString(r.denominator.String())
	return sb.String()
}

func (r Rational[T]) ToDecimalString() string {
	ten := r.getTen()
	qr := r.numerator.Abs().DivideByAndModulo(r.denominator)
	wholeInteger, remainder := qr[0], qr[1]
	var decimal strings.Builder
	if r.numerator.IsNegative() {
		decimal.WriteByte('-')
	}
	decimal.WriteString(wholeInteger.String())
	if !remainder.IsZero() {
		decimal.WriteByte('.')
	}
	remainderSeen := map[string]int{remainder.String(): decimal.Len()}
	startingCyclic := -1
	for !remainder.IsZero() && startingCyclic == -1 {
		remainder = remainder.Multiply(ten)
		qr = remainder.DivideByAndModulo(r.denominator)
		digit := qr[0]
		remainder = qr[1]
		decimal.WriteByte(digit.String()[0])
		key := remainder.String()
		if pos, ok := remainderSeen[key]; ok {
			startingCyclic = pos
		}
		remainderSeen[key] = decimal.Len()
	}
	result := decimal.String()
	if startingCyclic != -1 {
		result = result[:startingCyclic] + string(rune(cyclicBegin)) + result[startingCyclic:] + string(rune(cyclicEnd))
	}
	return result
}

func (r Rational[T]) ToTruncateDecimalString(roundTo int32) string {
	// Checked up front, like the math.MinInt32 guards in
	// ToRoundDecimalString and ToCeilDecimalString: math.MinInt32 has no
	// positive counterpart ("the minimum representable precision" isn't
	// representable), so there is nothing meaningful to truncate to and
	// every port rejects it. Checking it before negating anything (rather
	// than negating first and relying on how that happens to overflow)
	// means the rest of this function can just negate roundTo normally.
	if roundTo == math.MinInt32 {
		panic("cannot round to the minimum representable precision")
	}
	ten := r.getTen()
	qr := r.numerator.Abs().DivideByAndModulo(r.denominator)
	wholeInteger := qr[0]
	wholeIntegerStr := wholeInteger.String()
	sign := ""
	if r.numerator.IsNegative() {
		sign = "-"
	}
	negRoundTo := -roundTo
	if len(wholeIntegerStr) <= int(negRoundTo) {
		return sign + "0"
	} else if roundTo < 0 {
		shiftBase := ten.Pow(negRoundTo)
		result := wholeInteger.DivideBy(shiftBase).Multiply(shiftBase)
		return sign + result.String()
	} else if roundTo == 0 {
		return sign + wholeIntegerStr
	}
	remainder := qr[1]
	var decimal strings.Builder
	decimal.WriteString(sign)
	decimal.WriteString(wholeIntegerStr)
	if !remainder.IsZero() {
		decimal.WriteByte('.')
	}
	for i := int32(0); i < roundTo && !remainder.IsZero(); i++ {
		remainder = remainder.Multiply(ten)
		qr = remainder.DivideByAndModulo(r.denominator)
		digit := qr[0]
		remainder = qr[1]
		decimal.WriteByte(digit.String()[0])
	}
	return decimal.String()
}

func (r Rational[T]) ToRoundDecimalString(roundTo int32) string {
	if roundTo == math.MinInt32 {
		panic("cannot round to the minimum representable precision")
	}
	five := Rational[T]{r.getFive(), r.denominator.Pow(0)}
	ten := Rational[T]{r.getTen(), r.denominator.Pow(0)}
	shiftBase := ten.Pow(-roundTo - 1)
	delta := five.Multiply(shiftBase)
	var temp Rational[T]
	if r.IsNegative() {
		temp = r.Minus(delta)
	} else {
		temp = r.Plus(delta)
	}
	return temp.ToTruncateDecimalString(roundTo)
}

func (r Rational[T]) ToCeilDecimalString(roundTo int32) string {
	if r.IsNegative() {
		return "-" + r.Negate().ToFloorDecimalString(roundTo)
	}
	if roundTo == math.MinInt32 {
		panic("cannot round to the minimum representable precision")
	}
	ten := Rational[T]{r.getTen(), r.denominator.Pow(0)}
	if r.Multiply(ten.Pow(roundTo)).IsInteger() {
		return r.ToTruncateDecimalString(roundTo)
	}
	shiftBase := ten.Pow(-roundTo)
	temp := r.Plus(shiftBase)
	return temp.ToTruncateDecimalString(roundTo)
}

func (r Rational[T]) ToFloorDecimalString(roundTo int32) string {
	if r.IsNegative() {
		return "-" + r.Negate().ToCeilDecimalString(roundTo)
	}
	return r.ToTruncateDecimalString(roundTo)
}

func (r Rational[T]) IsInteger() bool  { return r.denominator.IsOne() }
func (r Rational[T]) IsZero() bool     { return r.numerator.IsZero() }
func (r Rational[T]) IsPositive() bool { return r.numerator.IsPositive() }
func (r Rational[T]) IsNegative() bool { return r.numerator.IsNegative() }

func (r Rational[T]) Negate() Rational[T] { return Rational[T]{r.numerator.Negate(), r.denominator} }
func (r Rational[T]) Abs() Rational[T]    { return Rational[T]{r.numerator.Abs(), r.denominator} }

func (r Rational[T]) Reciprocal() Rational[T] {
	if r.IsZero() {
		panic("cannot get the reciprocal of zero")
	}
	return rationalSignNormalized(r.denominator, r.numerator)
}

func (r Rational[T]) reciprocalUnsafe() Rational[T] {
	return rationalSignNormalized(r.denominator, r.numerator)
}

func (r Rational[T]) GetNumeratorAbs() T   { return r.numerator.Abs() }
func (r Rational[T]) GetDenominatorAbs() T { return r.denominator.Abs() }

func (r Rational[T]) Compare(other Rational[T]) int {
	gcd := r.denominator.Gcd(other.denominator)
	left := r.numerator.Multiply(other.denominator.DivideBy(gcd))
	right := other.numerator.Multiply(r.denominator.DivideBy(gcd))
	return left.Compare(right)
}

func (r Rational[T]) Equals(other Rational[T]) bool {
	return r.numerator.Equals(other.numerator) && r.denominator.Equals(other.denominator)
}

func (r Rational[T]) Plus(other Rational[T]) Rational[T] {
	gcd := r.denominator.Gcd(other.denominator)
	m1 := other.denominator.DivideBy(gcd)
	m2 := r.denominator.DivideBy(gcd)
	numerator := r.numerator.Multiply(m1).Plus(other.numerator.Multiply(m2))
	denominator := gcd.Multiply(m1).Multiply(m2)
	return rationalReduced(numerator, denominator)
}

func (r Rational[T]) Minus(other Rational[T]) Rational[T] {
	gcd := r.denominator.Gcd(other.denominator)
	m1 := other.denominator.DivideBy(gcd)
	m2 := r.denominator.DivideBy(gcd)
	numerator := r.numerator.Multiply(m1).Minus(other.numerator.Multiply(m2))
	denominator := gcd.Multiply(m1).Multiply(m2)
	return rationalReduced(numerator, denominator)
}

func (r Rational[T]) Multiply(other Rational[T]) Rational[T] {
	gcd1 := r.denominator.Gcd(other.numerator)
	gcd2 := other.denominator.Gcd(r.numerator)
	denominator := r.denominator.DivideBy(gcd1).Multiply(other.denominator.DivideBy(gcd2))
	numerator := r.numerator.DivideBy(gcd2).Multiply(other.numerator.DivideBy(gcd1))
	return Rational[T]{numerator, denominator}
}

func (r Rational[T]) DivideBy(other Rational[T]) Rational[T] {
	if other.IsZero() {
		panic("cannot divide by zero")
	}
	return r.Multiply(other.reciprocalUnsafe())
}

func (r Rational[T]) Pow(exponent int32) Rational[T] {
	if exponent < 0 {
		if r.IsZero() {
			panic("exponent cannot be negative for zero")
		}
		if exponent == math.MinInt32 {
			return r.reciprocalUnsafe().Pow(math.MaxInt32).Multiply(r.reciprocalUnsafe())
		}
		return r.reciprocalUnsafe().Pow(-exponent)
	} else if exponent == 0 {
		return r.getOne()
	} else if exponent == 1 {
		return r
	}
	result := r.getOne()
	base := r
	power := exponent
	for {
		if power&1 == 1 {
			result = result.Multiply(base)
		}
		power >>= 1
		if power > 0 {
			base = base.Multiply(base)
		} else {
			break
		}
	}
	return result
}
